File: packages/exochain-py/exochain/consent/bailment.py
"""Bailment consent proposals.

A *bailment* is a scoped, time-bounded delegation of custody from a bailor
to a bailee. :class:`BailmentBuilder` assembles and validates a proposal,
producing an immutable :class:`BailmentProposal` with a deterministic id.
"""

from __future__ import annotations

from hashlib import sha256
from struct import pack

from pydantic import BaseModel, ConfigDict

from ..errors import ConsentError
from ..types import Did, Hash256Hex

MAX_SAFE_INTEGER = 9_007_199_254_740_991
HLC_LOGICAL_MAX = 0xFFFF_FFFF


class HlcTimestamp(BaseModel):
    """Caller-supplied Hybrid Logical Clock timestamp."""

    model_config = ConfigDict(frozen=True)

    physical_ms: int
    logical: int


class BailmentProposal(BaseModel):
    """A validated, immutable bailment proposal."""

    model_config = ConfigDict(frozen=True)

    proposal_id: Hash256Hex
    bailor: Did
    bailee: Did
    scope: str
    duration_hours: int
    created_at: int
    created_at_logical: int


class BailmentBuilder:
    """Fluent builder for a :class:`BailmentProposal`.

    Example:
        >>> proposal = (
        ...     BailmentBuilder("did:exo:alice", "did:exo:bob")
        ...     .scope("read:medical-records")
        ...     .duration_hours(48)
        ...     .created_at_hlc(1_700_000_000_000, 0)
        ...     .build()
        ... )
    """

    def __init__(self, bailor: Did, bailee: Did) -> None:
        self._bailor: Did = bailor
        self._bailee: Did = bailee
        self._scope: str | None = None
        self._duration_hours: int | None = None
        self._created_at_hlc: HlcTimestamp | None = None

    def scope(self, scope: str) -> BailmentBuilder:
        """Set the scope descriptor for this bailment."""
        if not isinstance(scope, str) or not scope.strip():
            raise ConsentError("scope cannot be empty")
        self._scope = scope
        return self

    def duration_hours(self, hours: int) -> BailmentBuilder:
        """Set the duration of the bailment in hours."""
        self._duration_hours = _positive_safe_integer(hours, "duration_hours")
        return self

    def created_at_hlc(self, physical_ms: int, logical: int = 0) -> BailmentBuilder:
        """Set the caller-supplied HLC creation timestamp."""
        self._created_at_hlc = HlcTimestamp(
            physical_ms=_positive_safe_integer(physical_ms, "created_at physical_ms"),
            logical=_hlc_logical(logical),
        )
        return self

    def build(self) -> BailmentProposal:
        """Validate and produce a :class:`BailmentProposal`.

        The ``proposal_id`` is the hex-encoded SHA-256 of a canonical payload
        containing the bailor, bailee, scope, duration, and caller-supplied HLC.
        """
        if self._scope is None:
            raise ConsentError("scope is required")
        if self._duration_hours is None:
            raise ConsentError("duration_hours is required")
        if self._created_at_hlc is None:
            raise ConsentError("created_at_hlc is required")

        payload = _proposal_payload(
            self._bailor,
            self._bailee,
            self._scope,
            self._duration_hours,
            self._created_at_hlc,
        )
        proposal_id: Hash256Hex = sha256(payload).hexdigest()
        return BailmentProposal(
            proposal_id=proposal_id,
            bailor=self._bailor,
            bailee=self._bailee,
            scope=self._scope,
            duration_hours=self._duration_hours,
            created_at=self._created_at_hlc.physical_ms,
            created_at_logical=self._created_at_hlc.logical,
        )


def _proposal_payload(
    bailor: Did,
    bailee: Did,
    scope: str,
    duration_hours: int,
    created_at_hlc: HlcTimestamp,
) -> bytes:
    return b"\0".join(
        [
            bailor.encode("utf-8"),
            bailee.encode("utf-8"),
            scope.encode("utf-8"),
            pack("<Q", duration_hours),
            pack("<Q", created_at_hlc.physical_ms),
            pack("<I", created_at_hlc.logical),
        ]
    )


def _positive_safe_integer(value: int, field: str) -> int:
    if (
        not isinstance(value, int)
        or isinstance(value, bool)
        or value <= 0
        or value > MAX_SAFE_INTEGER
    ):
        raise ConsentError(f"{field} must be a positive safe integer")
    return value


def _hlc_logical(value: int) -> int:
    if (
        not isinstance(value, int)
        or isinstance(value, bool)
        or value < 0
        or value > HLC_LOGICAL_MAX
    ):
        raise ConsentError(
            f"created_at logical must be an integer between 0 and {HLC_LOGICAL_MAX}"
        )
    return value


__all__ = ["BailmentBuilder", "BailmentProposal", "HlcTimestamp"]

File: packages/exochain-sdk/src/consent/bailment.ts
/**
 * Consent bailment builder.
 *
 * A bailment represents scoped, time-bounded consent from a bailor to a
 * bailee. `BailmentBuilder` mirrors the Rust SDK's builder pattern and
 * produces a {@link BailmentProposal} whose `proposalId` is a content-
 * addressed SHA-256 over the canonical fields. Callers must provide the
 * HLC timestamp from their deterministic execution context; this SDK does
 * not read wall-clock time while constructing consent records.
 */

import { ConsentError } from '../errors.js';
import type { Did, Hash256 } from '../types.js';
import { validateDid } from '../identity/did.js';
import { sha256, bytesToHex } from '../crypto/hash.js';

const HLC_LOGICAL_MAX = 0xffff_ffff;

export interface HlcTimestamp {
  readonly physicalMs: number;
  readonly logical: number;
}

/** A validated bailment proposal. */
export interface BailmentProposal {
  readonly proposalId: Hash256;
  readonly bailor: Did;
  readonly bailee: Did;
  readonly scope: string;
  readonly durationHours: number;
  /** HLC physical milliseconds supplied by the caller. */
  readonly createdAt: number;
  /** HLC logical counter supplied by the caller. */
  readonly createdAtLogical: number;
}

/** Builder for a {@link BailmentProposal}. */
export class BailmentBuilder {
  readonly #bailor: Did;
  readonly #bailee: Did;
  #scope?: string;
  #durationHours?: number;
  #createdAtHlc?: HlcTimestamp;

  constructor(bailor: Did | string, bailee: Did | string) {
    this.#bailor = typeof bailor === 'string' ? validateDid(bailor) : bailor;
    this.#bailee = typeof bailee === 'string' ? validateDid(bailee) : bailee;
  }

  /** Set the scope string (e.g. `"data:medical"`). */
  public scope(scope: string): this {
    this.#scope = scope;
    return this;
  }

  /** Set the bailment duration in whole hours. */
  public durationHours(hours: number): this {
    this.#durationHours = hours;
    return this;
  }

  /** Set the caller-supplied HLC creation timestamp for this proposal. */
  public createdAtHlc(physicalMs: number, logical: number = 0): this {
    this.#createdAtHlc = validateHlcTimestamp(physicalMs, logical);
    return this;
  }

  /** Validate the builder state and produce a {@link BailmentProposal}. */
  public async build(): Promise<BailmentProposal> {
    if (this.#scope === undefined) {
      throw new ConsentError('scope is required');
    }
    if (this.#scope.length === 0) {
      throw new ConsentError('scope must be non-empty');
    }
    if (this.#durationHours === undefined) {
      throw new ConsentError('durationHours is required');
    }
    const durationHours = validatePositiveSafeInteger(
      this.#durationHours,
      'durationHours',
    );
    if (this.#createdAtHlc === undefined) {
      throw new ConsentError('createdAtHlc is required');
    }
    const proposalId = await computeProposalId(
      this.#bailor,
      this.#bailee,
      this.#scope,
      durationHours,
      this.#createdAtHlc,
    );

    return {
      proposalId,
      bailor: this.#bailor,
      bailee: this.#bailee,
      scope: this.#scope,
      durationHours,
      createdAt: this.#createdAtHlc.physicalMs,
      createdAtLogical: this.#createdAtHlc.logical,
    };
  }
}

/**
 * Deterministic content-addressed proposal ID. Layout uses NUL separators,
 * LE u64 duration/physical-ms values, and LE u32 logical counter.
 */
async function computeProposalId(
  bailor: Did,
  bailee: Did,
  scope: string,
  durationHours: number,
  createdAtHlc: HlcTimestamp,
): Promise<Hash256> {
  const enc = new TextEncoder();
  const bailorB = enc.encode(bailor);
  const baileeB = enc.encode(bailee);
  const scopeB = enc.encode(scope);
  const durationB = new Uint8Array(8);
  // Little-endian u64 encoding.
  const view = new DataView(durationB.buffer);
  view.setBigUint64(0, BigInt(durationHours), true);
  const createdPhysicalB = new Uint8Array(8);
  new DataView(createdPhysicalB.buffer).setBigUint64(
    0,
    BigInt(createdAtHlc.physicalMs),
    true,
  );
  const createdLogicalB = new Uint8Array(4);
  new DataView(createdLogicalB.buffer).setUint32(0, createdAtHlc.logical, true);

  const total =
    bailorB.length +
    1 +
    baileeB.length +
    1 +
    scopeB.length +
    1 +
    durationB.length +
    1 +
    createdPhysicalB.length +
    1 +
    createdLogicalB.length;
  const payload = new Uint8Array(total);
  let offset = 0;
  payload.set(bailorB, offset);
  offset += bailorB.length;
  payload[offset++] = 0;
  payload.set(baileeB, offset);
  offset += baileeB.length;
  payload[offset++] = 0;
  payload.set(scopeB, offset);
  offset += scopeB.length;
  payload[offset++] = 0;
  payload.set(durationB, offset);
  offset += durationB.length;
  payload[offset++] = 0;
  payload.set(createdPhysicalB, offset);
  offset += createdPhysicalB.length;
  payload[offset++] = 0;
  payload.set(createdLogicalB, offset);

  const digest = await sha256(payload);
  return bytesToHex(digest) as Hash256;
}

function validatePositiveSafeInteger(value: number, field: string): number {
  if (!Number.isSafeInteger(value) || value <= 0) {
    throw new ConsentError(`${field} must be a positive safe integer`);
  }
  return value;
}

function validateHlcTimestamp(physicalMs: number, logical: number): HlcTimestamp {
  if (!Number.isSafeInteger(physicalMs) || physicalMs <= 0) {
    throw new ConsentError(
      'createdAtHlc physicalMs must be a positive safe integer',
    );
  }
  if (
    !Number.isSafeInteger(logical) ||
    logical < 0 ||
    logical > HLC_LOGICAL_MAX
  ) {
    throw new ConsentError(
      `createdAtHlc logical must be an integer between 0 and ${HLC_LOGICAL_MAX}`,
    );
  }
  return { physicalMs, logical };
}
